src/wallet3/db_schema.cpp

```diff
@@ -420,5 +425,41 @@ namespace wallet
   {
     return prepared_get<int64_t>("SELECT output_count FROM metadata WHERE id=0;");
   }
+  void
+  WalletDB::save_keys(const std::string& spend_priv_str, const std::string& spend_pub_str, const std::string& view_priv_str, const std::string& view_pub_str)
+  {
+    std::string query = "SELECT spend_priv, spend_pub, view_priv, view_pub FROM metadata WHERE id=0";
+
+    auto st = prepared_st(query);
+    st->executeStep();
+    auto from_db = db::get<std::string, std::string, std::string, std::string>(st);
+    const auto loaded_spend_priv_str = std::get<0>(from_db);
+    const auto loaded_spend_pub_str = std::get<1>(from_db);
+    const auto loaded_view_priv_str = std::get<2>(from_db);
+    const auto loaded_view_pub_str = std::get<3>(from_db);
+
+    if ((loaded_spend_priv_str != "" && loaded_spend_priv_str != spend_priv_str) ||
+        (loaded_spend_pub_str != "" && loaded_spend_pub_str != spend_pub_str) ||
+        (loaded_view_priv_str != "" && loaded_view_priv_str != view_priv_str) ||
+        (loaded_view_pub_str != "" && loaded_view_pub_str != view_pub_str))
+          throw std::runtime_error("loaded keys do not match database file");
+
+    prepared_exec("UPDATE metadata SET spend_priv = ?, spend_pub = ?, view_priv = ?, view_pub = ? where id = 0;",
+        spend_priv_str,
+        spend_pub_str,
+        view_priv_str,
+        view_pub_str);
+
+  }
+
+  std::tuple<std::string, std::string, std::string, std::string>
+  WalletDB::load_keys()
```

Comment:
This should probably be a `std::tuple<crypto::secret_key, crypto::public_key, crypto::secret_key, crypto::public_key>`.

Getting BLOBs *out* of the database, though, is currently a bit awkward: we can get it out in a string, but then we have to copy that string into the crypto::whatever.

We can fix that by making `db::blob` and `db::blob_guts` types that work something like `blob_binder` but for output rather than input.  I think this will do it:
```C++
  // Wrapper for extracting BLOB values from a query without unnecessary copying.  This is intended
  // to be called via `db::get` such as:
  //
  //    auto [num, data] = db::get<int, blob>(st);
  //
  // The `.data` string view here will point to the BLOB data directly.  Note that this view remains
  // only value while the statement remains active, and so it must be used as needed immediately.
  // This also means that this is *unsuitable* for one-shot methods like `prepared_get` (which
  // finalize the statement before returning).
  struct blob {
      std::string_view data;
      blob(SQLite::Column&& col)
        : data{static_cast<const char*>(col.getBlob()), static_cast<size_t>(col.getBytes())} {}
  };

  // Takes a primitive struct from which we can directly initialize from the stored blob value.  The
  // type `T` must be usable with `make_from_guts`.  Unlike `blob` this value *is* suitable for use
  // in a one-shot method.
  template <typename T>
  struct blob_guts {
      T value;
      blob_guts(SQLite::Column&& col)
          : value{tools::make_from_guts<T>(blob(std::move(col)).data)} {}

      // Implicit rvalue-convertible to `T&&` so that you can use it somewhat transparently, for
      // example, allowing implicit conversion from a `std::tuple<..., blob_guts<T>>` into a
      // `std::tuple<..., T>`.
      operator T&&() && { return std::move(value); }
  };
```

and then the body of this function would become:
```C++
    return prepared_get<blob_guts<crypto::secret_key>, blob_guts<crypto::public_key>, blob_guts<crypto::secret_key>, blob_guts<crypto::public_key>>(
        "SELECT ....");
```